src/robot_dart/control/pd_control.cpp

```diff
@@ -26,9 +28,21 @@ namespace robot_dart {
             Eigen::VectorXd q = robot->positions(_controllable_dofs);
             Eigen::VectorXd dq = robot->velocities(_controllable_dofs);
 
+
+            Eigen::VectorXd error= Eigen::VectorXd::Zero(_control_dof);
+            for(int i=0;i<_control_dof;++i){
+              if(robot->dof(_controllable_dofs[i])->getJoint()->getType() ==
+                  dart::dynamics::RevoluteJoint::getStaticType()){
```

Comment:
`RevoluteJoint`s are not the only ones requiring extra care:

- `BallJoint` has a full 3D orientation and uses angle axis representation
- `EulerJoint` has a full 3D orientation and uses Euler axes representation
- `FreeJoint` represents a full 3D transformation and includes an orientation

Can you do those too? I can do them if you do not feel comfortable.

By the way, we should enable/disable this with a variable (passing through the constructor, having a default value to true: i.e., using them).

Reply:
This is something that I think should be handled in a different way:
since we are dealing with rotations in SE(3) with three angles I think that a method that avoids singularities of the rotation matrices is advisable.
I didn't find anything more concise than embedding the 3D rotations in quaternions and computing the error between them with the geodesic distance between them.
I am already using this in one of my modules.
If you are interested, we can try to implement this here.
Also, I am not familiar with these other types of joints and I am wondering how we can access the values of the angles:
I am used to seeing that a joint has a single value for is q.

Reply:
> This is something that I think should be handled in a different way:
> since we are dealing with rotations in SE(3) with three angles I think that a method that avoids singularities of the rotation matrices is advisable.
> I didn't find anything more concise than embedding the 3D rotations in quaternions and computing the error between them with the geodesic distance between them.

How about `dart::math::logMap(R_desired * R_current.transpose())`?

> Also, I am not familiar with these other types of joints and I am wondering how we can access the values of the angles:
> I am used to seeing that a joint has a single value for is q.

For each joint `getPositions()` gives a vector with the DOF values:

- For `BallJoint` this should be a 3x1 vector with angle-axis representation of the orientation (`dart::math::expMapRot(vector)` gives the 3x3 rotation matrix)
- For `EulerJoint` this should be a 3x1 vector with Euler angles representation (depending on the angle order [see [here](https://github.com/dartsim/dart/blob/main/dart/dynamics/EulerJoint.hpp#L95)], you can use `dart::math::eulerXXXToMatrix(vector)` [see [here](https://github.com/dartsim/dart/blob/main/dart/math/Geometry.hpp#L76)] to get the rotation matrix)
- For `FreeJoint` this should be a 6x1 vector. This can be converted to a full transformation (which then you can split into translation and orientation parts) using `dart::dynamics::FreeJoint::convertToTransform(vector)`. In the end the vector should look like [error_orientation_x, error_orientation_y, error_orientation_z, error_x, error_y, error_z].

If you are comfortable, please do this. If not, let me know and I can do this.

Reply:
I am not feeling comfortable with exponential and logarithmic maps over quaternions still.
Therefore I prefer not to dive into the implementation for the time being.

Reply:
No worries. Finish the rest and I'll do this.. ;)

Reply:
Forgot to say thanks for the work! Thanks a lot @dinies..

Reply:
Your welcome ! =)